specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
+{
+    // Hide the default download dialog.
+    wil::com_ptr<ICoreWebView2Settings> settings;
+    CHECK_FAILURE(m_webView->get_Settings(&settings));
+    CHECK_FAILURE(settings->put_ShouldDisplayDefaultDownloadDialog(FALSE));
+
+    // Register a handler for the `DownloadStarting` event.
+    // This example hides the default download dialog and shows a dialog box instead.
+    // The dialog box displays the default save path and allows the user to specify a different path.
+    // Selecting `OK` will save the download to the chosen path.
+    // Selecting `CANCEL` will cancel the download.
+    CHECK_FAILURE(m_webView->add_DownloadStarting(
+        Callback<ICoreWebView2DownloadStartingEventHandler>(
+            [this](
+                ICoreWebView2_3* sender,
+                ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT
+            {
+                // We avoid potential reentrancy from running a message loop in the download
+                // starting event handler by showing our download dialog via this lambda run
+                // asynchronously later outside of this event handler.
+                auto showDialog = [this, args]
+                {
+                    wil::com_ptr<ICoreWebView2Download> download;
+                    CHECK_FAILURE(args->get_Download(&download));
+
+                    INT64 expectedDownloadSizeInBytes = 0;
+                    CHECK_FAILURE(download->get_ExpectedDownloadSizeInBytes(
+                      &expectedDownloadSizeInBytes));
+
+                    wil::unique_cotaskmem_string uri;
+                    CHECK_FAILURE(download->get_Uri(&uri));
+
+                    wil::unique_cotaskmem_string mimeType;
+                    CHECK_FAILURE(download->get_MimeType(&mimeType));
+
+                    wil::unique_cotaskmem_string contentDisposition;
+                    CHECK_FAILURE(download->get_ContentDisposition(&contentDisposition));
+
+                    // Get the suggested path from the event args.
+                    wil::unique_cotaskmem_string resultFilePath;
+                    CHECK_FAILURE(args->get_ResultFilePath(&resultFilePath));
+
+                    std::wstring prompt =
+                        std::wstring(L"Enter result file path or select `OK` to use default path. ") +
+                        L"Select `Cancel` to cancel the download.";
+
+                    std::wstring description = std::wstring(L"URI: ") + uri.get() + L"\r\n" +
+                                                L"Mime type: " + mimeType.get() + L"\r\n" +
+                                                L"Expected Download size in bytes: " +
+                                                std::to_wstring(expectedDownloadSizeInBytes) +
+                                                L"\r\n";
+
+                    TextInputDialog dialog(
```

Comment:
Consider cases like multiple downloads happening with reentrancy during the dialog, or navigation during download event. At the very least we need comments in the samples saying how it works, and perhaps updates to the code to ensure it would work.

Reply:
Also a comment noting that a long running synchronous UI prompt or other blocking item on the UI thread can potentially block the WebView2 from doing anything.